BugFix: do not squash dockable mapper when visibility is toggled

SlySven opened the pull request:
The existing code was showing and hiding both the `dlgMapper` and, when it is contained within a floating/dockable QDockWidget, that container widget. This has the unfortunate side effect of causing that type of mapper to shrink to a minimal size when shown a second time.

It was confusing that the code to show and hide the mapper was actually contained within a method called `createMapper` - so that has been renamed to `toggleVisibilityOfOrCreateMapper` and the functionality - clearly dividable into two parts - has been refactored into two new methods that does one part; coincidentally one of them can now be called `createMapper`!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Release post highlight
Fixed a bug when using a floating/dockable mapper which would squash the map down to a minimal, (unusable without manual resizing) size if it was shown, hidden and then shown again.

SlySven committed:
BugFix: do not squash dockable mapper when visibility is toggled

The existing code was showing and hiding both the `dlgMapper` and, when it
is contained within a floating/dockable QDockWidget, that container widget.
This has the unfortunate side effect of causing that type of mapper to
shrink to a minimal size when shown a second time.

It was confusing that the code to show and hide the mapper was actually
contained within a method called `createMapper` - so that has been renamed
to `toggleVisibilityOfOrCreateMapper` and the functionality - clearly
dividable into two parts - has been refactored into two new methods that
does one part; coincidentally one of them can now be called `createMapper`!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
> Fixed a bug when using a floating/dockable mapper which would squash the map down to a minimal, (unusable without manual resizing) size if it was shown, hidden and then shown again.

Map windows won't stay shrunk after hiding and showing again.

Kebap commented:
Can't reproduce the bug on PTB therefore can't verify the fix. Please give more detailed instructions.

SlySven commented:
#### Steps to reproduce the issue / Reasons for adding feature:

1. Load a profile with a map but which does not create a map embedded into the `TMainConsole` or a sub-`TConsole`.
2. Click the mapper button to load and show the map - and if necessary undock it so it is floating.
3. Adjust the size to be reasonable and usable.
4. Click the mapper button to hide the map.
5. Click the mapper button a third time to reshow the map.

#### Error output / Expected result of feature

**Expected:** the map should be shown as it was before being hidden.

**Actual:** the map is squashed down to a minimal size (if the control panel at the bottom is set to be shown that will limit how narrow the *width* of window gets - if not the width will be even less and the height will probably be no more than enough to show the frame and just the control panel "show/hide" button!) and has to be manually resized to be useful...

Kebap commented:
> 1. Load a profile with a map but which does not create a map embedded into the `TMainConsole` or a sub-`TConsole`.

How would I know this? Don't understand what you're saying.

Kebap commented:
Oh, it seems I got lucky! 
When I was testing before, I used the big red X of my floating window in step 4.
Now with your instructions, I hid the map window with another click on the 🌍 icon.
Low and behold, showing the map again will shrink the window awkwardly.
Now to test the PR build

Kebap reviewed (approved):
Does what it's supposed to

vadi2 commented on src/Host.cpp:
```diff
@@ -3533,18 +3533,36 @@ bool Host::resetBackgroundImage(const QString &name)
 // case that may need to create a map widget before it loads/imports a
 // non-default (last saved map in profile's map directory.
-void Host::createMapper(bool loadDefaultMap)
+void Host::toggleVisibilityOfOrCreateMapper(const bool loadDefaultMap)
```
How about `createOrShowMapper`?

SlySven replied:
It also hides it though - I do agree that name is ugly - lemme think....

1. `toggleShownOrCreateMapper`
2. `showHideOrCreateMapper`

vadi2 replied:
Hmm, the semantics here mean nothing at all because they mean everything. Should it be refactored?

SlySven replied:
I'm not sure I get what you mean. :confused:

If you mean, "Why was it called *createMapper* when it does more than that?" I suspect it evolved to get to be like that...

vadi2 commented on src/Host.cpp:
```diff
@@ -3533,18 +3533,36 @@ bool Host::resetBackgroundImage(const QString &name)
+{
+    auto pMap = mpMap.data();
+    bool visStatus = mpMap->mpMapper->isVisible();
+    if (pMap->mpMapper->parentWidget()->inherits("QDockWidget")) {
```
This should be a method of its own, `if (pMap->mpMapper->floatingWidget) {`...

SlySven replied:
Tracing it back through the history I see it came from Chris in 2013 in 5549b3690b0cf5f3fd8099708318ccc9a768f049 but that looks as though it was something tagged onto an unrelated commit - probably to properly hide the floating/dockable widget introduced in 41e0cbbed58f518321328e236429ce4d7097c6d6 (Mudlet 1.2.9-pre4).

SlySven replied:
Done.

vadi2 reviewed (commented):
The fix looks good! But `toggleVisibilityOfOrCreateMapper()` is code smell, let's do something about it quick.

SlySven committed:
Revise: respond to peer review

Also add a `(bool) isFloatAndDockable()` method to the `dlgMapper` class
to refactor out the means to identify that instance of the class being
contained in a `QDockWidget`.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
> But `toggleVisibilityOfOrCreateMapper()` is code smell, let's do something about it quick.

That has been around for years. Let's push the fix for release before diving into refactoring

vadi2 reviewed (approved):
OK, make sure not to forget about this and move on

Kebap commented:
> OK, make sure not to forget about this and move on

You can open an issue for reference..
